File: EduLink.Domain/Strategies/CodigoPromocionalStrategy.cs
using EduLink.Domain.Entities;
using EduLink.Domain.Interfaces;
using System;

namespace EduLink.Domain.Strategies;

public class CodigoPromocionalStrategy : IPrecioStrategy
{
    private readonly string _codigoEsperado;
    private readonly decimal _descuentoPorcentual;
    private readonly decimal _descuentoFijo;

    public CodigoPromocionalStrategy(
        string codigoEsperado = "EDULINK10",
        decimal descuentoPorcentual = 0.10m,
        decimal descuentoFijo = 0m)
    {
        _codigoEsperado = codigoEsperado;
        _descuentoPorcentual = descuentoPorcentual;
        _descuentoFijo = descuentoFijo;
    }

    public decimal Calcular(decimal precioBase, Cliente cliente)
    {
        
        var codigoActivo = "EDULINK10"; // nombre

        if (codigoActivo == _codigoEsperado)
        {
            var descuento = precioBase * _descuentoPorcentual;
            if (_descuentoFijo > 0) descuento = Math.Min(descuento, _descuentoFijo);
            return Math.Max(0, precioBase - descuento);
        }

        return precioBase;
    }
}

File: app/domain/strategies/codigo_promocional_strategy.rb
require_relative '../interfaces/i_precio_strategy'

module Strategies
  class CodigoPromocionalStrategy
    include Interfaces::IPrecioStrategy

    def initialize(codigo_esperado: "EDULINK10", descuento_porcentual: 0.10, descuento_fijo: 0)
      @codigo_esperado = codigo_esperado
      @descuento_porcentual = descuento_porcentual
      @descuento_fijo = descuento_fijo
    end

    def calcular(precio_base, cliente)
      # Al final no creo que pongamos un formulario para codigos
      codigo_activo = "EDULINK10" # Simulación

      if codigo_activo == @codigo_esperado
        descuento = precio_base * @descuento_porcentual
        descuento = [@descuento_fijo, descuento].min if @descuento_fijo > 0
        [0, precio_base - descuento].max
      else
        precio_base
      end
    end
  end
end
